src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7615,6 +7619,74 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
     return S_OK;
 }
 
+HRESULT ProfToEEInterfaceImpl::EnumerateGCHeapObjects(ObjectCallback callback, void* callbackState)
+{
+    CONTRACTL
+    {
+        // Yay!
+        NOTHROW;
+
+        // Suspending EE is contracted to trigger GC through ThreadSuspend::SuspendAllThreads
+        GC_TRIGGERS;
+
+        // Yay!
+        MODE_ANY;
+
+        // Suspending EE will acquire the ThreadStore Lock via ThreadSuspend::LockThreadStore.
+        CAN_TAKE_LOCK;
+
+        // Yay!
+        EE_THREAD_NOT_REQUIRED;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers,
+        (LF_CORPROF,
+        LL_INFO1000,
+        "**PROF: EnumerateGCHeapObjects.\n"));
+
+    if (callback == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+
+    if (!g_fEEStarted)
+    {
+        return CORPROF_E_RUNTIME_UNINITIALIZED;
+    }
+
+    if (!ThreadSuspend::SysIsSuspendInProgress() && (ThreadSuspend::GetSuspensionThread() == 0))
```

Comment:
Are you trying to only suspend if the profiler has not already suspended the runtime? I worry about this check because it doesn't distinguish between a profiler induced suspension and a random suspension for some other reason.

Also, the `RestartEE` call below isn't in an if statement so if this does not suspend the runtime, it will try to restart the runtime in an unbalanced way.

Is there a reason we can't just unconditionally suspend the runtime here?

Reply:
I added logic to suspend/resume EE in the same manner that `ProfToEEInterfaceImpl::SuspendRuntime` and `ProfToEEInterfaceImpl::ResumeRuntime` do it. I haven't yet delved into the details of whether a double `ThreadSuspend::SuspendEE` before a restart would cause any issues, @noahfalk/@jkotas would y'all happen to know? 

Should this API not proceed if there was a non-profiler induced suspension (since whatever suspended EE might try to resume EE before this API finishes)? We could check for a previous profiler-induced suspension with `fProfilerRequestedRuntimeSuspend`, but then a similar question pops up, whether that previous profiler-induced suspension is already paired with a resumption.

Would it make sense to require profilers to handle EE Suspension/Resumption outside of this EnumerateGCHeapObjects API rather than building it into the logic?
e.g.
```cpp
pCorProfilerInfo->SuspendRuntime();
...
pCorProfilerInfo->EnumerateGCHeapObjects(heap_walk_fn, &state);
...
pCorProfilerInfo->ResumeRuntime();
```

Reply:
> double ThreadSuspend::SuspendEE

I do not think double ThreadSuspend::SuspendEE is going to work correctly. You can prove or disprove it by adding tests.

Reply:
Both double `ThreadSuspend::SuspendEE` and `ThreadSuspend::RestartEE` will fail assertions, added logic to only invoke `ThreadSuspend::RestartEE` in this API if `ThreadSuspend::SuspendEE` had been invoked in the API.

Reply:
Give that this API has `EE_THREAD_NOT_REQUIRED` contract, it is expected to work on background threads that have not joined the runtime. If this is called on a background thread, there may be suspension running on the foreground thread, `ThreadSuspend::SysIsSuspendInProgress()` returns true and we will skip our own suspension here. However, nothing guarantees that the suspension that was performed by the foreground thread is going to last as long as this method runs. In extreme case, the runtime may be resumed and running by the time we hit the next line that sets `g_profControlBlock.fGCInProgress = TRUE;`.

I think the contract of the method needs to be made more stringent to avoid this situation; or the implementation should be fixed to handle this situation.

Reply:
I've been answering some questions in other communication channels piece-meal but since it still sounded like there was ongoing uncertainty I wanted to offer a more definitive suggestion.

I think this would give us good behavior:
```
bool ownEESuspension = false;
bool suspendedByThisThread = (ThreadSuspend::GetSuspensionThread() == GetThreadNullOK());
if (suspendedByThisThread && !g_profControlBlock.fProfilerRequestedRuntimeSuspend)
{
    // This thread is responsible for suspending the runtime so we can't block
    // waiting for the runtime to resume. However it wasn't the profiler call that did
    // the suspend so we don't know what state the GC heap is in. Other threads also might 
    // be modifying it concurrently. In the future more analysis or coordination with other
    // suspenders might let us narrow the scope of this error condition, but we have no need
    // to do this now.
    return CORPROF_E_SUSPENSION_IN_PROGRESS;
}
else if(suspendedByThisThread && g_profControlBlock.fProfilerRequestedRuntimeSuspend)
{
    // This thread previously invoked ICorProfiler::SuspendRuntime(). Our caller
    // has the responsibility to resume the runtime no earlier than when this API returns
    // and to preserve the GC heap in a consistent state. We should avoid invoking 
    // SuspendEE/ResumeEE again in this function because those APIs do not support 
    // re-entrant suspends.
}
else
{
    _ASSERTE(!suspendedByThisThread);
    // Its possible some background threads are suspending and resuming the runtime
    // concurrently. We need to suspend the runtime on this thread to be certain the heap
    // stays in a walkable state for the duration that we need it to. Our call to
    // SuspendEE() may race with other threads by design and this thread may block
    // arbitrarily long inside SuspendEE() for other threads to complete their own
    // suspensions.
    g_profControlBlock.fProfilerRequestedRuntimeSuspend = TRUE;
    ThreadSuspend::SuspendEE(ThreadSuspend::SUSPEND_REASON::SUSPEND_FOR_PROFILER);
    ownEESuspension = TRUE;
}
```

Reply:
Thanks for the suggestion! Added it